### PR TITLE
[ao] Added constant channel check to Outlier Detector

### PR DESCRIPTION
**Pull request by vspenubarthi**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81369
* #81382
* #81252
* #81251
* #81250
* __->__ #81249
* #81174

Summary: The current Outlier detector does a good job of finding whether
data distributions passing through layers have outliers. However,
suppose we have a completely constant channel. The outlier detector
would not detect it as an outlier, but that is still something we want
to highlight because a constant channel usually is a result of a bad
configuration or something really wrong with the data.

To address this there are two additions to the outlier detector that
this commit makes:
- The first is to add whether there are any constant batches at all and
let the user know in the text report
- The second is to let the user know the number of total constant
batches found for each channel, so they can figure out if there are any
unnecessary channels present.

The exisiting outlier detector tests were modified to do a quick check
for this feature.

Test Plan: python test/test_quantization.py TestFxDetectOutliers

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 1]
```diff
@@ -1079,11 +1085,13 @@ def _calculate_outlier_info(
             percentile_ratios (torch.Tensor): The average percentile_ratios per channel calculated by the observer
             counted_batches (torch.Tensor): The number of batches used for average calculation per tensor
             total_batches (int): The total number of batches that passed through observer in this epoch
+            constant_channels (torch.Tensor): The counts for constant batches per channel
 
         Returns a dictionary mapping:
             "outliers_detected" : list of bools per channel that are true if it is considered an outlier
             "sufficient_batches": if o_r was >= statistical_threshold:
                 where o_r = counted_batches / total_batches
+            "constant_channels": if non_constant_per_channel != total_batches for each channel
```
Comment (andrewor14) [inline torch/ao/quantization/fx/_model_report/detector.py:1094]: What is `non_constant_per_channel`? I couldn't find any references to this anywhere

Comment (andrewor14) [reply torch/ao/quantization/fx/_model_report/detector.py:1094]: Also is this a list of bools or a list of counts? Should clarify in comments

File: torch/ao/quantization/fx/_model_report/detector.py [hunk 2]
```diff
@@ -1164,6 +1181,8 @@ def _generate_info_dict(self, model: GraphModule) -> Dict[str, Dict]:
                     self.NUM_BATCHES_KEY: num_batches,
                     self.OUTLIER_KEY: outlier_calcs[self.OUTLIER_KEY],
                     self.SUFFICIENT_BATCHES_KEY: outlier_calcs[self.SUFFICIENT_BATCHES_KEY],
+                    self.CONSTANT_CHANNEL_KEY: outlier_calcs[self.CONSTANT_CHANNEL_KEY],
+                    self.CONSTANT_COUNTS_KEY: channel_batch_cnts,
```
Comment (andrewor14) [inline torch/ao/quantization/fx/_model_report/detector.py:1185]: Do we need both of these? It seems if `num_constant_batches` == 0 then `has_constant_batches` will be False? Correct me if I'm misunderstanding something but the latter seems redundant.

**Review by andrewor14 — APPROVED** [review]

File: torch/ao/quantization/fx/_model_report/detector.py (hunk 2 above)
Comment (vspenubarthi) [reply torch/ao/quantization/fx/_model_report/detector.py:1185]: you are right that if `num_constant_batches == 0` then has_constant_batches is False, i just had done this to make the logic of generating the text report easier.

However, this is redundant data, so I will make this fix by removing the boolean value.

**Pull request closed without merging**